### ISSUE TITLE
Cookie not writable exception: Wrong cookie/settings path on Windows, bad directory separator

### ISSUE THREAD

**Issue by Kozzy1055** [mgp25/Instagram-API#1590]
### Before submitting an issue make sure you have:
- [x] [Searched](https://github.com/mgp25/Instagram-API/search?type=Issues) the bugtracker for similar issues including **closed** ones
- [x] [Read the FAQ](https://github.com/mgp25/Instagram-API/wiki/FAQ)
- [x] [Read the wiki](https://github.com/mgp25/Instagram-API/wiki)
- [x] [Reviewed the examples](https://github.com/mgp25/Instagram-API/tree/master/examples)
- [x] [Installed the api using ``composer``](https://github.com/mgp25/Instagram-API#installation)

### Purpose of your issue?
- [x] Bug report (encountered problems/errors)
- [ ] Feature request (request for new functionality)
- [ ] Question
- [ ] Other

### Describe your issue

Hi there!

On Windows server your code is not working. Keep getting exception about cookie file is not writable.

When you write the settings and cookie file, with `Utils::atomicWrite` you using a safe mechanism for writing:create new temp file then rename it. But at the rename it is always fails on Windows servers.

The bug is in the Windows internal rename function, it cant accept `/` mark as directory separator which you using.

**Solution:** Replace your `/` with system dependant `DIRECTORY_SEPARATOR` constant, its works for all systems.

Here are the lines you need to check out:

`Settings/Storage/File.php`

around line 53 in `public function openLocation(`

```php
:Constants::SRC_DIR.DIRECTORY_SEPARATOR.'..'.DIRECTORY_SEPARATOR.'sessions'.DIRECTORY_SEPARATOR);
```

around line 329 in `private function _generateUserPaths(`

```php
$userFolder = $this->_baseFolder.$username.DIRECTORY_SEPARATOR;
```

Could you fix these please?
Thank you!

Keep up the good job!

**Comment by SteveJobzniak** [comment]
Thanks a lot for the detailed investigation. We all really appreciate it!

Turns out that you have found a bug in PHP. They claim to always do translation of forward slashes to the correct OS path, so that forward slashes are always safe. But they’ve forgotten to do system-independent translation of `/` to `\` on Windows systems during `rename()`. Thye jsut pass it raw to the system. Bad, evil PHP! ;-)

I will investigate all of our paths that are part of rename() calls and fix them in the next release.

**Comment by valga** [comment]
@SteveJobzniak PHP has a [test](https://github.com/php/php-src/blob/6053987bc27e8dede37f437193a5cad448f99bce/ext/standard/tests/file/rename_variation-win32.phpt) for it since 2010. And it works for me with PHP 5.6.30 on Windows:
```php
var_dump(PHP_OS);
file_put_contents('../test/before', 'omg');
var_dump(rename('../test/before', '../test/after'));
var_dump(file_get_contents('../test/after'));
```
Output:
```
string(5) "WINNT"
bool(true)
string(3) "omg"
```

**Comment by SteveJobzniak** [comment]
@valga Damn. Nice discovery.

I wonder what's wrong then... Perhaps the problem is the mixed path. Something like `C:\Users\Heh\Desktop\project\vendor\mgp25\instagram-php/sessions/Username/wahtever.dat` may be confusing it because of mixed `\` and `/` ?

Could one of you please test that? I don't have Windows.

**Comment by valga** [comment]
@SteveJobzniak 
```php
file_put_contents('../test\mixed', 'slashes');
var_dump(rename('..\test/mixed', '../test\result'));
var_dump(file_get_contents('..\test/result'));
```
```
string(5) "WINNT"
bool(true)
string(7) "slashes"
```

**Comment by SteveJobzniak** [comment]
Damn that's weird... @Kozzy1055 please help investigate what's really wrong here... Seems there's no problem with our `rename()` and our `/` slashes...  You are the first user who has had a problem (tons of windows users run it successfully) so please help us figure out what's wrong. :-O Perhaps something in your particular file-path is confusing PHP in very special cases...

**Comment by valga** [comment]
@SteveJobzniak PHP for Windows just replaces all `/` with `\` in its path normalizer function:
https://github.com/php/php-src/blob/07e8bdd8c2bc6a88ab54d41b3ebd3685cb421257/win32/ioutil.c#L524-L527

So it works with both back and forward slashes.

I think the error comes from some permissions stuff.

**Comment by SteveJobzniak** [comment]
Ah perfect, so the path normalizer runs exactly as I expected. That makes me happy, because I totally HATE `DIRECTORY_SEPARATOR`, it's so annoying to write... And we can't alias it to something shorter since definitions are global and could conflict with the user's own code (ie `define('DS', DIRECTORY_SEPARATOR);` = NOT SAFE, since the person using our library may have made their own 'DS' that we then overwrite).

Compare:

```php
require __DIR__.'/../vendor/autoload.php'
// vs
require __DIR__.DIRECTORY_SEPARATOR.'..'.DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'autoload.php'
```

So I really hope that the error is filesystem-permission related.

I think it is. Because as I said, I realized we have tons of Windows users who all use the default `File` settings backend and for the past half-year (since the new settings system was released) we have had ZERO tickets about problems writing the cookies.

@Kozzy1055 You say you're on a windows SERVER. Those have stricter folder-permission controls than regular windows. Please verify the permissions. Also check your PHP version (5.6, or preferably 7.0+).

It's weird that you were able to make the code work by changing to the DIRECTORY_SEPARATOR. Since PHP's path-normalize which valga links to always replaces `/` with `\` on Windows...

**Comment by SteveJobzniak** [comment]
@Kozzy1055 Here's the relevant code that valga linked to. As you can see, PHP's `php_win32_ioutil_normalize_path_w()` function translates EVERY path-string as follows:

```c
	while (NULL != (pos = wcschr(idx, PHP_WIN32_IOUTIL_FW_SLASHW)) && (size_t)(idx - *buf) <= len) {
		*pos = PHP_WIN32_IOUTIL_DEFAULT_SLASHW;
		idx = pos++;
	}
```

In other words, it's like a `$path = str_replace('/', DIRECTORY_SEPARATOR, $path);` built-into PHP itself. It replaces EVERY forward-slash it finds in the path, with a backslash instead.

So our paths, which we generate with `bleh\bloh/bleh\` will look like `bleh/bloh/bleh` when it finally gets sent by PHP to the final Windows OS `rename()` syscall.

Even PHP themselves have unit tests since 2010 to verify that their `rename()` properly remaps the slashes as intended and works: https://github.com/php/php-src/blob/6053987bc27e8dede37f437193a5cad448f99bce/ext/standard/tests/file/rename_variation-win32.phpt.

And valga had success.

In other words... no, we don't need to edit our library to `DIRECTORY_SEPARATOR` (that horrible thing). And thank god for that, because it's truly a horrible way to write paths. ;-)

Anyway, please let us know what the actual problem is. Probably a folder-permission issue.

**Comment by SteveJobzniak** [comment]
One more thing: Nowhere in original post claims that you actually tried the DIRECTORY_SEPARATOR rewrite and that it worked. You only said that you got warnings that the file is not writable.

So I am starting to think you never tried changing our code to `DIRECTORY_SEPARATOR`. And instead just looked at the Exception "permissions error" with the `\bleh/bloh` mixed-style path and just assumed that was the problem?

I think we can close this soon. ;-)

**Comment by mgp25** [comment]
You can follow error in the following issue: https://github.com/mgp25/Instagram-API/issues/1591
